### PR TITLE
Feature/runtime updates lite

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
~I'm looking at why the tests are failing, though the implementation seems correct~
...[later]...
It seems fixed now. egeoip:start() was failing silently because some required apps (from app.src) were not started

File: src/egeoip_db_updater.erl [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+-spec handle_info(Info :: reload
+                        | term(),
+                  State :: state())
+      -> {noreply, NewState :: state()}.
+handle_info(reload_db, #state{ dbpath = DbPath, db = Db } = State) ->
+    NewState
+        = case httpc:request(get, {DbPath, []}, [{timeout, ?REFRESH_TIMEOUT_MS}], []) of
```
Comment (g-andrade) [inline src/egeoip_db_updater.erl]: Do you think it would be worth it to go for conditional requests? (The 'if-modified-since' stuff.)

File: test/egeoip_tests.erl [hunk 2]
```diff
@@ -59,17 +62,27 @@ egeoip() ->
...
+        = case application:get_env(egeoip, dbfile) of
+              {ok, city} -> <<"CA">>;
+              _ -> <<>>
+          end,
     #geoip{country_code = "US",
            country_code3 = "USA",
            country_name = "United States",
-           region = <<"CA">>,
+           region = Region,
            _ = _} = R,
```
Comment (g-andrade) [inline test/egeoip_tests.erl:75]: I would suggest always using the [assert macros](http://erlang.org/doc/apps/eunit/chapter.html), as they tend to give more helpful output in case of failure.

File: src/egeoip.app.src [hunk 3]
```diff
@@ -3,6 +3,12 @@
   {vsn,"1.1"},
   {modules,[]},
   {registered, []},
-  {env, [{dbfile, city}]},
+  {env, [
+      {dbfile, city}%,
+      % dbpath may be left undefined, so as to not run the self-updater
+      %   but if defined, it must be a dat.gz file
+      %{dbpath, "https://geolite.maxmind.com/download/geoip/database/GeoLiteCity.dat.gz"}
+      % or e.g. "https://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz"
```
Comment (g-andrade) [inline src/egeoip.app.src]: Do we plan on using maxmind's URLs? If so, I fear we might risk throttling. Perhaps we can mirror them instead.

File: src/egeoip_db_updater.erl (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply src/egeoip_db_updater.erl]: I searched for info regarding the request headers, but found none. If that part of the request is not "official" I don't know to what extent can we trust it (or implement on top of)

File: test/egeoip_tests.erl (hunk 2 above)
Comment (paulo-ferraz-oliveira) [reply test/egeoip_tests.erl:75]: In what sense? [I just made sure the tests didn't fail with already-expected app.config values, here]

File: src/egeoip.app.src (hunk 3 above)
Comment (paulo-ferraz-oliveira) [reply src/egeoip.app.src]: That is still do be decided, I guess. Wherever we put them (I think we might have to consult MaxMind about unexpected use like mirroring, ...) the option will adapt to that (as long as it's gz). But I wanted an example for app.config and chose the one I could best remember :D

In any case, we're downloading a database every 24 hours. Would that provoke throttling?

File: test/egeoip_tests.erl (hunk 2 above)
Comment (paulo-ferraz-oliveira) [reply test/egeoip_tests.erl:75]: You mean change the already-existing tests to not do pattern-matching? And instead use something like
`?assertMatch(#geoip{country_code = "US",
           country_code3 = "USA",
           country_name = "United States",
           region = Region1,
           _ = _}, R1)`

File: src/egeoip_db_updater.erl (hunk 1 above)
Comment (g-andrade) [reply src/egeoip_db_updater.erl]: http://erlang.org/doc/man/httpc.html#request-4
If I read this correctly, the second argument, which is of type `request()`, is defined as `{url(), headers()} | {url(), headers(), content_type(), body()}`. So, in the case of the call above, the headers would go into the empty list sitting to the right of the URL.

File: test/egeoip_tests.erl (hunk 2 above)
Comment (g-andrade) [reply test/egeoip_tests.erl:75]: Exactly.

File: src/egeoip.app.src (hunk 3 above)
Comment (g-andrade) [reply src/egeoip.app.src]: Probably not.

Comment (fabiomsouto) [reply src/egeoip.app.src]: There's no official statements of throttling; if it becomes an issue we can host it on our end anyway.

Edit: After doing some cURL calls I see it's cloudflare hosted, so I don't think it'll be a problem.

File: src/egeoip.app.src [hunk 4]
```diff
@@ -3,6 +3,12 @@
   {vsn,"1.1"},
   {modules,[]},
   {registered, []},
-  {env, [{dbfile, city}]},
+  {env, [
+      {dbfile, city}%,
+      % dbpath may be left undefined, so as to not run the self-updater
```
Comment (fabiomsouto) [inline src/egeoip.app.src]: Would/should a filesystem path be supported too? Might be interesting for those cases where we might want to update the file outside of the application itself.

File: src/egeoip_db_updater.erl (hunk 1 above)
Comment (fabiomsouto) [reply src/egeoip_db_updater.erl]: Sounds good. Their hosts will happily comply (or cloudflare hosts, for that matter).

For a date before the last modified date:

```bash
⇒  curl -I https://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz -H 'If-Modified-Since: Tue, 04 Jul 2017 14:47:51 GMT'
HTTP/2 200
date: Mon, 31 Jul 2017 17:14:03 GMT
content-type: application/octet-stream
content-length: 640135
set-cookie: __cfduid=ddf44648f70fbd724b17657930c2758441501521243; expires=Tue, 31-Jul-18 17:14:03 GMT; path=/; domain=.maxmind.com; HttpOnly
last-modified: Tue, 04 Jul 2017 14:47:52 GMT
etag: "595baa98-9c487"
cf-cache-status: HIT
expires: Mon, 31 Jul 2017 21:14:03 GMT
cache-control: public, max-age=14400
accept-ranges: bytes
server: cloudflare-nginx
cf-ray: 38723119bf846206-LIS
```

For a date after the last modified date:
```bash
⇒  curl -I https://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz -H 'If-Modified-Since: Tue, 04 Jul 2017 14:47:52 GMT'
HTTP/2 304
date: Mon, 31 Jul 2017 17:13:58 GMT
set-cookie: __cfduid=d300a40a861d11fdabf78166ea98c2c8d1501521238; expires=Tue, 31-Jul-18 17:13:58 GMT; path=/; domain=.maxmind.com; HttpOnly
last-modified: Tue, 04 Jul 2017 14:47:52 GMT
etag: "595baa98-9c487"
cf-cache-status: HIT
expires: Mon, 31 Jul 2017 21:13:58 GMT
cache-control: public, max-age=14400
server: cloudflare-nginx
cf-ray: 387230ff4f146212-LIS
```

What would the time be though?

Comment (paulo-ferraz-oliveira) [reply src/egeoip_db_updater.erl]: "http://erlang.org/doc/man/httpc.html#request-4
If I read this correctly, the second argument, which is of type request(), is defined as {url(), headers()} | {url(), headers(), content_type(), body()}. So, in the case of the call above, the headers would go into the empty list sitting to the right of the URL."

Guilherme, I was talking about the response :D I don't know if that header is respected, in this case. What I meant was "I searched for info regarding the request headers**[' handling]**, but found none."

Comment (paulo-ferraz-oliveira) [inline src/egeoip_db_updater.erl]: Fábio, I don't think I understand the question "What would the time be though?", but it seems related to mine. We would ask only if something is if-modified-since the latest 24 hours, Guilherme?

File: test/egeoip_tests.erl (hunk 2 above)
Comment (paulo-ferraz-oliveira) [reply test/egeoip_tests.erl:75]: I didn't want to change the tests; your request and/or comment is valid, but outside the scope of the proposed changes.

File: src/egeoip.app.src (hunk 4 above)
Comment (paulo-ferraz-oliveira) [inline src/egeoip.app.src]: This PR only handles remote files (in format .gz), which is our use case and current concern. If we have a use case for local files, sure, we can do another PR and support that.

Comment (paulo-ferraz-oliveira) [reply src/egeoip.app.src]: ~Will do~
Done

File: src/egeoip.app.src (hunk 3 above)
Comment (paulo-ferraz-oliveira) [reply src/egeoip.app.src]: Nothing to be done, then

File: test/egeoip_tests.erl (hunk 2 above)
Comment (paulo-ferraz-oliveira) [reply test/egeoip_tests.erl:75]: Nothing to be done, here

File: src/egeoip_db_updater.erl (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply src/egeoip_db_updater.erl]: I don't think this is an issue, for now, as the 12 hour-interval should have minimal impact in the system (except for the one peak, for a 600 KB download). Nothing will be done regarding this

File: src/egeoip.app.src (hunk 4 above)
Comment (paulo-ferraz-oliveira) [inline src/egeoip.app.src]: Done

**Comment by paulo-ferraz-oliveira** [comment]
~Tests are failing in OTP R16B. If I have time I'll look into it~
...[later]...
Erlang/OTP R16B does not have application:ensure_all_started

**Review by fabiomsouto — APPROVED** [review]

**Review by gnsr — APPROVED** [review]
